File: packages/guards/src/prompt-length.ts
import type { Guard, GuardResult, GuardContext } from 'open-guardrail-core';

interface PromptLengthOptions { action: 'block' | 'warn'; maxChars?: number; maxWords?: number; maxLines?: number; }

export function promptLength(options: PromptLengthOptions): Guard {
  return { name: 'prompt-length', version: '0.1.0', description: 'Validate input prompt length (chars/words/lines)', category: 'format', supportedStages: ['input'],
    async check(text: string, _ctx: GuardContext): Promise<GuardResult> {
      const start = performance.now(); const violations: string[] = [];
      if (options.maxChars && text.length > options.maxChars) violations.push(`${text.length} chars (max: ${options.maxChars})`);
      if (options.maxWords) { const words = text.split(/\s+/).filter(w => w.length > 0).length; if (words > options.maxWords) violations.push(`${words} words (max: ${options.maxWords})`); }
      if (options.maxLines) { const lines = text.split('\n').length; if (lines > options.maxLines) violations.push(`${lines} lines (max: ${options.maxLines})`); }
      const triggered = violations.length > 0;
      return { guardName: 'prompt-length', passed: !triggered, action: triggered ? options.action : 'allow',
        message: triggered ? `Prompt too long: ${violations.join(', ')}` : undefined,
        latencyMs: Math.round(performance.now() - start),
        details: triggered ? { violations, reason: 'Input prompt exceeds length limits' } : undefined,
      };
    },
  };
}

File: python/open_guardrail/guards/prompt_length.py
"""Check prompt character length bounds."""
from __future__ import annotations

import time

from open_guardrail.core import GuardResult


class _PromptLength:
    def __init__(
        self,
        *,
        action: str = "block",
        max_length: int = 10000,
        min_length: int = 1,
    ) -> None:
        self.name = "prompt-length"
        self.action = action
        self._max = max_length
        self._min = min_length

    def check(
        self, text: str, stage: str = "input"
    ) -> GuardResult:
        start = time.perf_counter()
        length = len(text)
        triggered = (
            length > self._max
            or length < self._min
        )
        elapsed = (
            time.perf_counter() - start
        ) * 1000

        if not triggered:
            return GuardResult(
                guard_name="prompt-length",
                passed=True,
                action="allow",
                latency_ms=round(elapsed, 2),
            )

        if length > self._max:
            msg = (
                f"Prompt too long: {length}"
                f" > {self._max} chars"
            )
        else:
            msg = (
                f"Prompt too short: {length}"
                f" < {self._min} chars"
            )

        return GuardResult(
            guard_name="prompt-length",
            passed=False,
            action=self.action,
            message=msg,
            latency_ms=round(elapsed, 2),
            details={
                "length": length,
                "max_length": self._max,
                "min_length": self._min,
                "reason": msg,
            },
        )


def prompt_length(
    *,
    action: str = "block",
    max_length: int = 10000,
    min_length: int = 1,
) -> _PromptLength:
    return _PromptLength(
        action=action,
        max_length=max_length,
        min_length=min_length,
    )
